src/profile-logic/transforms.js

```diff
@@ -905,6 +937,76 @@ const FUNC_MATCHES = {
   },
 };
 
+export function collapseFunctionSubtree(
+  thread: Thread,
+  funcToCollapse: IndexIntoFuncTable
+): Thread {
+  const { stackTable, frameTable, samples } = thread;
+  const oldStackToNewStack: Map<
+    IndexIntoStackTable | null,
+    IndexIntoStackTable | null
+  > = new Map();
+  oldStackToNewStack.set(null, null);
+  const collapsedStacks = new Set();
+  const newStackTable = {
+    length: 0,
+    prefix: [],
+    frame: [],
+  };
+
+  for (let stackIndex = 0; stackIndex < stackTable.length; stackIndex++) {
+    const prefix = stackTable.prefix[stackIndex];
+    const frameIndex = stackTable.frame[stackIndex];
+    const funcIndex = frameTable.func[frameIndex];
+
+    if (
+      // The previous stack was collapsed, this one is collapsed too.
+      collapsedStacks.has(prefix)
+    ) {
+      // Only remember that this stack is collapsed.
+      const newPrefixStackIndex = oldStackToNewStack.get(prefix);
+      if (newPrefixStackIndex === undefined) {
+        throw new Error('newPrefixStackIndex cannot be undefined');
+      }
+      oldStackToNewStack.set(stackIndex, newPrefixStackIndex);
```

Comment:
Can you please add a comment explaining that because a stack is collapsed into one line, several index will point to the same "new stack index"?

Reply:
Ok, I tried to clarify a bit.